### DISCUSSION TITLE
Ways of reducing memory usage?

### DISCUSSION THREAD

**Discussion by refi64** [geldata/gel#3384] [Q&A]
Hello! I'm playing around with EdgeDB a bit, but I noticed the minimum memory usage is a bit...well, high. In particular, my projects tend to lean towards "you could self-host this on a cheap server", and I'm trying to see how low I can get EdgeDB's RAM usage in order to fit into that use case.

Out of the box, `docker stats` shows memory usage like this:

```
CONTAINER ID   NAME       CPU %     MEM USAGE / LIMIT     MEM %     NET I/O   BLOCK I/O         PIDS
1cf7f9cb69ba   edgedb     0.00%     644.4MiB / 23.22GiB   2.71%     0B / 0B   3.81MB / 0B       12
c30911765555   postgres   0.00%     298MiB / 23.22GiB     1.25%     0B / 0B   11.4MB / 1.02GB   9
```

(I ran postgres separately so I could more easily see how much EdgeDB itself was using.)

After some playing around, I noticed I can change the number of the compiler processes, and EdgeDB already only uses 1 if the memory is <1GB. Passing `--compiler-pool-size 1` gives memory usage that looks like this:

```
CONTAINER ID   NAME       CPU %     MEM USAGE / LIMIT     MEM %     NET I/O   BLOCK I/O         PIDS
c30911765555   postgres   0.00%     112.3MiB / 23.22GiB   0.47%     0B / 0B   11.4MB / 1.02GB   8
82c5745460e7   edgedb     0.00%     405.1MiB / 23.22GiB   1.70%     0B / 0B   13.1MB / 0B       11
```

Now, this is *lower*, but still kind of high. I *did* notice there still seem to be 2 compiler processes running:

```
$ docker exec -it edgedb ps -Awwo rss,command
  RSS COMMAND
196904 /usr/lib/x86_64-linux-gnu/edgedb-server-1-rc3/bin/python3.10 /usr/bin/edgedb-server --bind-address=0.0.0.0 --bind-address=:: --port=5656 --backend-dsn=postgresql://postgres:postgres@localhost:5432/ --runstate-dir=/run/edgedb --tls-cert-mode=generate_self_signed --tls-cert-file=/etc/ssl/edgedb/edbtlscert.pem --tls-key-file=/etc/ssl/edgedb/edbprivkey.pem --default-auth-method=Trust --http-endpoint-security=optional --compiler-pool-size 1 --emit-server-status=/tmp/tmp.uNJGBlj2BO
    0 [sleep] <defunct>
    0 [bash] <defunct>
    0 [bash] <defunct>
161816 /usr/lib/x86_64-linux-gnu/edgedb-server-1-rc3/bin/python3.10 -m edb.server.compiler_pool.worker --sockname /run/edgedb/4t79ox65/ipc --numproc 1 --version-serial 1
172992 /usr/lib/x86_64-linux-gnu/edgedb-server-1-rc3/bin/python3.10 -m edb.server.compiler_pool.worker --sockname /run/edgedb/4t79ox65/ipc --numproc 1 --version-serial 1
 1272 ps -Awwo rss,command
```

Slightly easier-to-read from smem:

```
$ sudo $(which smem) -aktP '^\S+edgedb\S+python3'     
  PID User    Command                                                                     Swap     USS     PSS     RSS 
90827 polkitd /usr/lib/x86_64-linux-gnu/edgedb-server-1-rc3/bin/python3.10 -m edb.serv       0   59.9M  108.1M  158.0M 
90828 polkitd /usr/lib/x86_64-linux-gnu/edgedb-server-1-rc3/bin/python3.10 -m edb.serv       0   77.4M  122.3M  168.9M 
90676 polkitd /usr/lib/x86_64-linux-gnu/edgedb-server-1-rc3/bin/python3.10 /usr/bin/ed       0  180.9M  185.8M  192.3M 
-----------------------------------------------------------------------------------------------------------------------
    3 1                                                                                      0  318.2M  416.2M  519.3M 
```

Not sure if one of these is just a forked off helper, or one of them is a monitoring process, etc.

So I guess my question is:

- Is it possible to reduce the memory usage of this any further?
- Is there a reason there still appears to be two compiler processes running?

**Comment by fantix** [comment]
I can answer one of the 2 questions:

> Not sure if one of these is just a forked off helper, or one of them is a monitoring process, etc.

You're right about both - the parent compiler worker process is a hot fork helper, but it's also a monitoring process which forks itself as soon as the number of workers is lower than what's demanded.

**Comment by elprans** [comment]
The only possible further reduction is adding a single-process mode where the compiler runs in the same process as I/O, so there is just one EdgeDB process overall.

**Reply by 1st1** [reply]
Yeah, we can run the compiler in a thread, so that there can happen *some* concurrent IO while it's compiling.
